### PR TITLE
🏡 Homepage re-design: Docs Week Q2 edition

### PR DESCRIPTION
**Pull request by olyism**
Implement the homepage redesign. See screenshot below:

Screen Shot 2023-04-20 at 12 03 05 PM

**We re-designed the homepage to:**
- Improve homepage information architecture and content usability:
  - Easier way-finding for content pertinent to the user's needs — for both new and experienced users
  - Remove visual clutter and reduce page length
- Improve emotional design to provide a more delightful homepage experience

**Details on implementations:**
1. Foundation work:
  1. Delete existing Homepage frontend code 🔥
  2. Refactor and simplify page containers on all page types
  3. Re-organize `erb` template folder structure
  4. Prepare common CSS utilities
2. Add header navigation to the homepage
3. Build and style components for the sections
  - Hero banner
  - Products
  - Popular guides
  - References
4. Clean ups and fixes:
  - Use correct Sass variables for box shadow and lightest text color
  - Spacing tweaks

Also, responsive small screen screenshot

Screen Shot 2023-04-20 at 12 12 58 PM

Screen Shot 2023-04-20 at 12 13 53 PM

### CODE REVIEW

**Comment by olyism** [comment]
Manually tested on Browserstack

### Edge, Windows

Screen Shot 2023-04-20 at 12 16 40 PM

### Safari, iOS

Screen Shot 2023-04-20 at 12 21 33 PM

### Samsung browser

Screen Shot 2023-04-20 at 12 22 54 PM

Also, search looks good on mobile device

Screen Shot 2023-04-20 at 12 22 35 PM

File: app/assets/stylesheets/base/_mixins.scss
Comment (olyism) [inline app/assets/stylesheets/base/_mixins.scss:1]: A new shared mixin to create CSS containers. I used this to constrain the width of the homepage's one-column layout.

I then refactored the existing `.PageContainer` styles to extend from this mixin.

File: app/assets/stylesheets/base/_variables.scss [hunk 1]
```diff
@@ -61,7 +61,7 @@ $slate-900: #151921;
 // color aliases
 $color-aliases: (
   "text-base": $charcoal-700,
-  "text-lightest": $charcoal-400,
```
Comment (olyism) [inline app/assets/stylesheets/base/_variables.scss:64]: Small fix to correct our lightest a11y text color.

File: app/assets/stylesheets/base/_variables.scss [hunk 2]
```diff
@@ -111,10 +111,12 @@ $gutter: 32px;
 $max-w: (
   "outer-container": 1600px,
   "container": 1440px,
-  "prose": 720px,
+  "narrow-container": 960px,
+  "prose": 752px,
```
Comment (olyism) [inline app/assets/stylesheets/base/_variables.scss:115]: Adjusted `prose` to work with the `container` mixin's horizontal padding.

File: app/assets/stylesheets/components/_button.scss [hunk 3]
```diff
@@ -3,6 +3,14 @@
   font-size: $font-size-template;
 }
 
+@mixin button-white-text {
```
Comment (olyism) [inline app/assets/stylesheets/components/_button.scss:6]: Ensures buttons with white text remains white even if it's nested within `.TextContent`

**Comment by dannymidnight** [comment]
Thoughts on full-width hero? 

Screenshot 2023-04-20 at 2 09 56 pm

File: app/views/homepage/_references.html.erb [hunk 4]
```diff
@@ -0,0 +1,101 @@
+<%
+  references = [
+    {
+      title: "Pipeline",
+      links: [
+        {
+          title: "Step types",
+          link: "/docs/pipelines/command-step"
```
Comment (mbelton-buildkite) [inline app/views/homepage/_references.html.erb:8]: If you merge main, you should be able to link to the overview: `/docs/pipelines/step-reference`

File: app/views/homepage/_references.html.erb [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+          title: "Configuration settings",
+          link: "/docs/agent/v3/configuration"
+        },
+        {
+          title: "Lifecycle hooks",
+          link: "/docs/agent/v3/hooks"
+        },
+        {
+          title: "CLI commands",
+          link: "/docs/agent/v3/cli-start"
```
Comment (mbelton-buildkite) [inline app/views/homepage/_references.html.erb:38]: If you merge main, you should be able to link to the overview: `/docs/agent/v3/cli-reference`

File: app/views/homepage/_references.html.erb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          title: "Webhooks",
+          link: "/docs/apis/webhooks"
+        }
+      ]
+    },
+    {
+      title: "Test Analytics",
+      links: [
+        {
+          title: "JSON test results data",
```
Comment (mbelton-buildkite) [inline app/views/homepage/_references.html.erb:67]: Let's shorten this slightly:

```suggestion
          title: "JSON test results",
```

**Review by mbelton-buildkite — COMMENTED** [review]
Just a couple of tweaks, but looks amazing! 👏🏻

**Review by mbelton-buildkite — APPROVED** [review]
Looks good to me! ✨

File: app/assets/stylesheets/components/_button.scss (hunk 3 above)
Comment (dannymidnight) [reply app/assets/stylesheets/components/_button.scss:6]: This doesn't feel quite right having a mixin include a class override. It seems like `.TextContent` is being a little too specific so perhaps that selector is worth revisiting.

File: app/assets/stylesheets/pages/_docs.scss [hunk 7]
```diff
@@ -1,13 +1,33 @@
-$color-green: rgb(3, 165, 98);
+@mixin margin-left-sidebar {
+  margin-left: $sidebar-width + ($gutter / 2);
+}
 
-.Page {
```
Comment (dannymidnight) [inline app/assets/stylesheets/pages/_docs.scss:3]: It looks like you've brought back a bunch of styles that I had removed. We might need to revisit some of the merge conflicts.

File: app/controllers/pages_controller.rb [hunk 8]
```diff
@@ -3,6 +3,9 @@ class PagesController < ApplicationController
   layout :layout_by_path
 
   def index
+    @nav = default_nav
+    puts @nav
```
Comment (dannymidnight) [inline app/controllers/pages_controller.rb:7]: This can be removed.

File: app/models/nav.rb [hunk 9]
```diff
@@ -12,9 +12,7 @@ def current_item_root
 
   # Returns the current nav item
   def current_item
-    item = route_map[path.sub("/docs/", "")]
-    raise ActionController::RoutingError.new("Missing navigation for #{path}") unless item
```
Comment (dannymidnight) [inline app/models/nav.rb:16]: I found this error useful for handling orphaned pages. I think we should instead add a condition for the home page separately.

e.g

```
return nil if path =="/"
```

File: app/views/homepage/_references.html.erb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+          link: "/docs/test-analytics/ci-environments"
+        }
+      ]
+    }
+  ]
+%>
+
+<section class="References">
+  <div class="container container-narrow">
+    <h2 class="HomePage__section-heading">APIs and references</h2>
```
Comment (dannymidnight) [inline app/views/homepage/_references.html.erb:81]: I'm not loving the mixed levels of BEM and surgical classes as it makes the styles hard to follow.

Since we've got SASS mixins at our disposal could you do something like the following?

```scss
.References { 
  @import container("narrow")
}

.References__Header {
  @include section-header;
}
```

Maybe the surgical classes (`container`) is OK but definitely having `Homepage__section-heading` nested under `References` is what I would consider to be a BEM violation

**Comment by olyism** [comment]
> Thoughts on full-width hero?

Ha, good thing to debate. My personal preference is constrain hero banner. Especially on large or extra wide monitors it starts to look like a strip. We also use a maximum container width that's consistent with the app. Lastly, a fixed container width helps to keep the visual attention focused.

Here's a full-width example:

Screen Shot 2023-04-21 at 11 47 54 AM

On my largest monitor 👆

Screen Shot 2023-04-21 at 11 48 08 AM

On an ultra wide monitor 👆

----

For this PR I'm leaning towards a consistent container width for the hero, but happy to chat about aesthetics for the next round of improvements.

**Review by dannymidnight — COMMENTED** [review]
It's all looking schmick.

Left some feedback and I think we should have a chat about some of the styles that have been reverted in the last PR to understand if that was on purpose or not 😁

**Comment by dannymidnight** [comment]
> Ha, good thing to debate. My personal preference is constrain hero banner. Especially on large or extra wide monitors it starts to look like a strip. We also use a maximum container width that's consistent with the app. Lastly, a fixed container width helps to keep the visual attention focused.

OK I said full-width, but I didn't actually mean fuuuuuull width. More the width of the nav container.

File: app/assets/stylesheets/components/_button.scss (hunk 3 above)
Comment (olyism) [reply app/assets/stylesheets/components/_button.scss:6]: Yeah I agree, I don't like the specificity problem with `.TextContent`. We could possibly use `:not` selector. I had a try but was getting out-of-scope for this PR.

I'll raise a backlog ticket to revisit the best way to handle standard markdown v.s. non-standard elements `.TextContent` styling.

File: app/views/homepage/_references.html.erb (hunk 10 above)
Comment (olyism) [reply app/views/homepage/_references.html.erb:81]: Personally, the developer experience of using BEM in Docs isn't helpful with styling and communicating the component structure. It's a pattern I've inherited and I'd be keen to move away from this. We should chat about other ways to work with the frontend.

File: app/assets/stylesheets/pages/_docs.scss (hunk 7 above)
Comment (olyism) [reply app/assets/stylesheets/pages/_docs.scss:3]: Ah yes, this was on purpose. I reworked this to achieve a few things (forgot to leave comments in the commit):

1. Moved spacing to header from `.Page` to `.PageContainer`
2. I don't think we should @extend `.TextContent` to `.Page`.

Might be easier to sync up to chat about this one.

File: app/models/nav.rb (hunk 9 above)
Comment (olyism) [reply app/models/nav.rb:16]: Ah I see. I just pushed up a fix [here](https://github.com/buildkite/docs/pull/2054/files#diff-1788a142b31ad28644197424968d69a9d0b8ffca46a37605dfdaaf74345fd409R15)

```
return nil if path == "/docs"
```

File: app/views/homepage/_references.html.erb (hunk 10 above)
Comment (olyism) [reply app/views/homepage/_references.html.erb]: I've pushed up a commit to use the Sass mixin fa627346668f78e25a8ecc26a35790b867372854

File: app/controllers/pages_controller.rb (hunk 8 above)
Comment (olyism) [reply app/controllers/pages_controller.rb]: Thanks, good catch. I've pushed up the fix.

File: app/assets/stylesheets/pages/_docs.scss (hunk 7 above)
Comment (olyism) [reply app/assets/stylesheets/pages/_docs.scss:3]: We paired and resolved this.

**Review by dannymidnight — APPROVED** [review]
👍
